spec/features/script_spec.rb

```diff
@@ -248,27 +248,55 @@
   end
 end
 
-# describe "count_and_sort.rake" do
-#   it "outputs list of sorted numbers with a count", points: 1 do
-#     path_to_file = Rails.root + "lib/input_files/descriptive_statistics_numbers.txt"
-#     string_input = open(path_to_file).read.chomp
-#     numbers = string_input.gsub(",", "").split.map(&:to_f)
-#
-#     sorted_numbers = numbers.sort
-#     count = numbers.count
-#
-#     regex_string = sorted_numbers.inject("") {|num, str| "#{Regexp.escape(num)}.*#{str}" }
-#     sorted_numbers_regex = Regexp.new(regex_string)
-#     ap(sorted_numbers_regex)
-#
-#     rake_task_output = capture_rake_task_output(task_name: 'count_and_sort')
-#
-#     ap(rake_task_output)
-#
-#     expect(rake_task_output.index(sorted_numbers_regex)).to be_truthy
-#
-#   end
-# end
+describe "count_and_sort.rake" do
+  it "outputs list of sorted numbers with a count", points: 1 do
+    path_to_file = Rails.root + "lib/input_files/descriptive_statistics_numbers.txt"
+    string_input = open(path_to_file).read.chomp
+    numbers = string_input.gsub(",", "").split.map(&:to_f)
+
+    sorted_numbers = numbers.sort
+    count = numbers.count
+
+    rake_task_output = capture_rake_task_output(task_name: 'count_and_sort')
+    numbers_index_end = count + 1
+    count_index = count + 3
+    sorted_numbers_start = count + 6
+    sorted_numbers_end = sorted_numbers_start + count - 1
+
+    expect(rake_task_output.split("\n").count).to eql 27
+    rake_task_output.split("\n").each_with_index do |line, index|
+      answer_line = ""
+      case index
+      when 0
+        answer_line = "Your numbers:"
+      when 1
+        answer_line = "["
+      when 2..numbers_index_end
+        answer_line = "[" + (index - 2).to_s + "] " + numbers[(index - 2)].to_s
+      when numbers_index_end + 1
+        answer_line = "]"
+      when count_index
+        answer_line = "Count: #{count}"
+      when count_index + 1
+        answer_line = "Sorted Numbers:"
+      when sorted_numbers_start - 1
+        answer_line = "["
+      when sorted_numbers_start..sorted_numbers_end
+        answer_line =  "[" + (index - count - 6).to_s + "] " + sorted_numbers[(index - count - 6)].to_s
+      when sorted_numbers_end + 1
+        answer_line = "]"
+      end
+
+      if answer_line.size > 1 && answer_line.at(0) == "["
+        expect(line.index(answer_line.split.at(0))).to be_truthy
+        expect(line.index(answer_line.split.at(1))).to be_truthy
+      else
+        expect(line.index(answer_line)).to be_truthy
```

Comment:
It took me a long time to wrap my head around this assertion, but I _think_ I get it!

When we're writing Capybara specs, we usually just test for the presence of content on the page. Could we do something similar here, and just look for the presence of the entire output (perhaps as a heredoc) that we want in stdout?

[This link](https://www.eliotsykes.com/test-rails-rake-tasks-with-rspec) seems to have some guidance on adding some helper methods that might come in handy, like

```ruby
  it "logs to stdout" do
    expect { task.execute }.to output("Sending invoices\n").to_stdout
  end
```

I could be totally misunderstanding though because I'm not able to clone the code at present

Reply:
The issue I struggled with is that the output of a rake task has a bunch of  weird characters mixed in that I don't fully understand. They sometimes change, I think? Here's what the output looks like:
```
"\e[0;33m\"Your numbers:\"\e[0m\n\e[0;33m\"[\"\e[0m\n\e[0;33m\"     [0] 13.0,\"\e[0m\n\e[0;33m\"     [1] 32.0,\"\e[0m\n\e[0;33m\"     [2] 1.0,\"\e[0m\n\e[0;33m\"     [3] 4.0,\"\e[0m\n\e[0;33m\"     [4] 11.0,\"\e[0m\n\e[0;33m\"     [5] 17.0,\"\e[0m\n\e[0;33m\"     [6] 9.0,\"\e[0m\n\e[0;33m\"     [7] 10.0,\"\e[0m\n\e[0;33m\"     [8] 2.0,\"\e[0m\n\e[0;33m\"     [9] 4.0,\"\e[0m\n\e[0;33m\"]\"\e[0m\n\e[0;33m\"Count: 10\"\e[0m\n\e[0;33m\"Sorted Numbers:\"\e[0m\n\e[0;33m\"[\"\e[0m\n\e[0;33m\"     [0] 1.0,\"\e[0m\n\e[0;33m\"     [1] 2.0,\"\e[0m\n\e[0;33m\"     [2] 4.0,\"\e[0m\n\e[0;33m\"     [3] 4.0,\"\e[0m\n\e[0;33m\"     [4] 9.0,\"\e[0m\n\e[0;33m\"     [5] 10.0,\"\e[0m\n\e[0;33m\"     [6] 11.0,\"\e[0m\n\e[0;33m\"     [7] 13.0,\"\e[0m\n\e[0;33m\"     [8] 17.0,\"\e[0m\n\e[0;33m\"     [9] 32.0,\"\e[0m\n\e[0;33m\"]\"\e[0m\n"  
```
I had looked through the link you suggested before and still ran into this weirdness with the output that the article doesn't seem to acknowledge.

Also for this particular test, order is important, at least for printing out the sorted numbers. I wasn't sure how to test the order besides doing a line by line assertion.

Reply:
I think those are ansi escape codes for the colorization that awesome_print adds.

A cursory google turned up this, try it out and see if it helps?

https://github.com/piotrmurach/strings-ansi#23-sanitize

Reply:
> Also for this particular test, order is important, at least for printing out the sorted numbers. I wasn't sure how to test the order besides doing a line by line assertion.

As far as the order, I thought if you expected the presence of the _entire_ output that we want (as one giant heredoc) then that would take care of that?

Reply:
You're definitely right that the order _would_ be tested if we compared it to a heredoc. I think I was considering that off the table since I wasn't sure how to parse the ansi characters. I'll look into the article you linked and try some more.

Reply:
Alternatively could we even use awesome_print itself to produce the string of output, including the ansi escapes, that we are expecting?

Reply:
@jelaniwoods Ah ha I think I found a method that might come in handy, `#awesome_inspect`:

```
Running via Spring preloader in process 3641
Loading development environment (Rails 5.1.6)
[1] pry(main)> ap [1, 2, 3]
[
    [0] 1,
    [1] 2,
    [2] 3
]
=> nil
[2] pry(main)> [1, 2, 3].awesome_inspect
=> "[\n    \e[1;37m[0] \e[0m1,\n    \e[1;37m[1] \e[0m2,\n    \e[1;37m[2] \e[0m3\n]"
```

Reply:
From what I recall of these specs, most of the copy in the output we don't really care about, e.g. "The original numbers were..."

So we can just assert that, somewhere in stdout, they awesome printed the correct array — or whatever value. Sort of like we just look for the correct content anywhere in an HTML page.

What do you think?

Reply:
@raghubetina `awesome_inspect` is super helpful! I was having trouble getting the input to match with a heredoc.

Reply:
I'm assuming the test would fail if they didn't use `awesome print` and tried to print each line with a loop. Should that be the expected behavior?

Reply:
We'll have to think through how we want to structure rake task assignments in the future. Continue teaching them to use `ap` rather than `puts`? If so, then yes they should be using whatever we show them, not trying to replicate `ap` behavior with `puts`, which is a huge waste of effort. We should make the instructions emphasize not to try and do that.

Reply:
I didn't mean necessarily that students would use `puts` instead of `ap`, just that they don't always realize you can print an Array with `ap`. But emphasizing `ap`, maybe in the ruby cheatsheet guide if it's not already there, seems like a good idea.